File: src/com/rian/difficultycalculator/math/Precision.java
package com.rian.difficultycalculator.math;

/**
 * Precision utilities.
 */
public final class Precision {
    /**
     * The default epsilon for all <code>float</code> values.
     */
    public static final float FLOAT_EPSILON = 1e-3f;

    /**
     * The default epsilon for all <code>double</code> values.
     */
    public static final double DOUBLE_EPSILON = 1e-7;

    /**
     * Checks if two numbers are equal with a given tolerance.
     *
     * @param value1 The first number.
     * @param value2 The second number.
     */
    public static boolean almostEqualsNumber(float value1, float value2) {
        return almostEqualsNumber(value1, value2, FLOAT_EPSILON);
    }

    /**
     * Checks if two numbers are equal with a given tolerance.
     *
     * @param value1 The first number.
     * @param value2 The second number.
     * @param acceptableDifference The acceptable difference as threshold.
     */
    public static boolean almostEqualsNumber(float value1, float value2, float acceptableDifference) {
        return almostEqualsNumber(value1, value2, (double) acceptableDifference);
    }

    /**
     * Checks if two numbers are equal with a given tolerance.
     *
     * @param value1 The first number.
     * @param value2 The second number.
     */
    public static boolean almostEqualsNumber(double value1, double value2) {
        return almostEqualsNumber(value1, value2, DOUBLE_EPSILON);
    }

    /**
     * Checks if two numbers are equal with a given tolerance.
     *
     * @param value1 The first number.
     * @param value2 The second number.
     */
    public static boolean almostEqualsNumber(double value1, double value2, double acceptableDifference) {
        return Math.abs(value1 - value2) <= acceptableDifference;
    }
}

File: src/com/rian/osu/math/Precision.kt
package com.rian.osu.math

import kotlin.math.abs

/**
 * Precision utilities.
 */
object Precision {
    /**
     * The default epsilon for all [Float] values.
     */
    const val FLOAT_EPSILON = 1e-3f

    /**
     * The default epsilon for all [Double] values.
     */
    const val DOUBLE_EPSILON = 1e-7

    /**
     * Checks if a [Float] is definitely greater than another [Float] with a given tolerance.
     *
     * @param value1 The first [Float].
     * @param value2 The second [Float].
     * @param acceptableDifference The acceptable difference. Defaults to [FLOAT_EPSILON].
     * @return Whether [value1] is definitely greater than [value2].
     */
    @JvmStatic
    @JvmOverloads
    fun definitelyBigger(value1: Float, value2: Float, acceptableDifference: Float = FLOAT_EPSILON) =
        value1 - acceptableDifference > value2

    /**
     * Checks if a [Double] is definitely greater than another [Double] with a given tolerance.
     *
     * @param value1 The first [Double].
     * @param value2 The second [Double].
     * @param acceptableDifference The acceptable difference. Defaults to [DOUBLE_EPSILON].
     * @return Whether [value1] is definitely greater than [value2].
     */
    @JvmStatic
    @JvmOverloads
    fun definitelyBigger(value1: Double, value2: Double, acceptableDifference: Double = DOUBLE_EPSILON) =
        value1 - acceptableDifference > value2

    /**
     * Checks if two numbers are equal with a given tolerance.
     *
     * @param value1 The first number.
     * @param value2 The second number.
     * @param acceptableDifference The acceptable difference as threshold.
     */
    @JvmStatic
    @JvmOverloads
    fun almostEquals(value1: Float, value2: Float, acceptableDifference: Float = FLOAT_EPSILON) =
        abs(value1 - value2) <= acceptableDifference

    /**
     * Checks if two numbers are equal with a given tolerance.
     *
     * @param value1 The first number.
     * @param value2 The second number.
     * @param acceptableDifference The acceptable difference as threshold.
     */
    @JvmStatic
    @JvmOverloads
    fun almostEquals(value1: Double, value2: Double, acceptableDifference: Double = DOUBLE_EPSILON) =
        abs(value1 - value2) <= acceptableDifference
}
